### PR TITLE
Test that queued query does not have data field

### PR DESCRIPTION
**Pull request by wendigo**
This is undocumented property of a protocol that some clients rely upon.

Test this requirement explicitly.

### CODE REVIEW

**Comment by nineinchnick** [comment]
> This is undocumented property of a protocol that some clients rely upon.

Do we want to document it in https://github.com/trinodb/trino/blob/master/docs/src/main/sphinx/develop/client-protocol.md ?

**Comment by wendigo** [comment]
@nineinchnick added

**Comment by martint** [comment]
As a general comment, clients should not rely on the query state to make decisions. That field is just for informational purposes only. For example, the query is not done until the next uri is null, regardless of what the state field says.

I don’t recall of there are any guarantees regarding the data field being empty while the state indicates “queued”, although I could imagine potential future cases that make it legit for a query to be marked as queues while data is in the output buffer (e.g, the query gets parked and re-queued to make way for other queries).

cc @electrum

**Review by hashhar — CHANGES_REQUESTED** [review]
IMO the clients should be fixed rather than encoding this arbitrary requirement into the engine.

**Comment by electrum** [comment]
Which clients rely on this?

**Comment by xsgao-github** [comment]
> > This is undocumented property of a protocol that some clients rely upon.
> 
> Do we want to document it in [`master`/docs/src/main/sphinx/develop/client-protocol.md](https://github.com/trinodb/trino/blob/master/docs/src/main/sphinx/develop/client-protocol.md?rgh-link-date=2024-02-29T12%3A02%3A17Z) ?

Do we have a JSON schema to set the binding between client and server? If so, we should publish it. Words is not enough to set expectations from both sides.

ODBC driver works when `data` attribute is totally missing from the response. But fails when `data: null` is returned. Per @wendigo

**Comment by findepi** [comment]
> As a general comment, clients should not rely on the query state to make decisions. That field is just for informational purposes only. For example, the query is not done until the next uri is null, regardless of what the state field says.

agreed


> Which clients rely on this?


AFAIU, all [the ODBC drivers](https://trino.io/ecosystem/client#odbc) have a problem when JSON response contains `data: null` and work well when `data` key is omitted (or present when response carries data).

The https://trino.io/docs/current/develop/client-protocol.html#important-queryresults-attributes should explicitly state whether `data: null`  is valid part of response, even though current Trino version doesn't emit that.
If we specify it is not, this property of protocol should be tested. Why? Because we use jackson everywhere, and so our existing tests cannot verifies protocol properties that jackson does not discern. At the same time, AFAIU, we do not want jackson's behavior to be implicit part of the protocol specification.

File: client/trino-client/src/main/java/io/trino/client/QueryResults.java [hunk 1]
```diff
@@ -141,6 +142,7 @@ public List<Column> getColumns()
     @Nullable
     @JsonProperty
     @Override
+    @JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (findepi) [inline client/trino-client/src/main/java/io/trino/client/QueryResults.java:145]: PR description suggests we're adding a test only, here we modify the core classes.
Is this "just a test" or "a fix with a test"?

File: docs/src/main/sphinx/develop/client-protocol.md [hunk 2]
```diff
@@ -77,7 +77,8 @@ class `io.trino.client.QueryResults` in module `trino-client` in the
 * - `data`
   - The `data` attribute contains a list of the rows returned by the query
     request. Each row is itself a list that holds values of the columns in the
-    row, in the order specified by the `columns` attribute.
+    row, in the order specified by the `columns` attribute. If there is no data,
+    attribute is omitted.
```
Comment (findepi) [inline docs/src/main/sphinx/develop/client-protocol.md:81]: is -> should be


if we were describing current implementation, present simple tense would be appropriate.
for the protocol specification, imperative form is more appropriate

cc @mosabua

Comment (mosabua) [reply docs/src/main/sphinx/develop/client-protocol.md:81]: fair

**Pull request closed without merging**